rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,315 @@
+at a future stage, but is out of scope for this RFC.
+
+Likewise, reloading the enrichment data is likely to be needed, but will not be
+covered here. For the first implementation of this RFC, reloading the data will
+require Vector to be restarted.
+
+## Motivation
+
+Users would like to enrich events flowing through Vector with extra data
+provided by a CSV file.
+
+## Doc-level Proposal
+
+### Resources
+
+To represent the CSV file we have a new top level configuration option.
+
+```
+[resources.csv_file]
+  type = "csv"
+  path = "\path_to_csv"
+  delimiter = ","
+```
+
+The fields available for this section are:
+
+*type*
+
+The type of data that the resource represents. Currently only a csv file is
+supported, but this may expand in future to handle other file formats and
+database connections.
+
+*path*
+
+The path to the csv file, either an absolute path or one relative to the current
+working directory.
+
+*delimiter*
+
+The delimiter used in the csv file to separate fields. Defaults to `","`.
+
+*title_row*
+
+If true, it assumes the first row in the csv file contains column names. If
+false, columns are named according to their numerical index.
+
+Note, this is likely to be where any encryption parameters such as the key are
+specified when it comes time to implementing encryption.
+
+The initial implementation will only be supporting CSV files as a resource. It
+is anticipated that future work will expand the available resources to include
+other file types as well as databases.
+
+Like the services that sources and sinks integrate with, enrichment "resources"
+will likely need administrator approval and setup. Typically data is enriched
+from some sort of shared company "resource" that end users will likely not have
+access to. Admin will likely want to restrict access to resources to approved
+pipelines, especially if sensitive information is contained.
+
+The resources will need to be integrated with `vector validate`, which would
+ensure the resource exists and is correctly formatted.
+
+### Schema
+
+For the CSV Resource all columns will be considered to be Strings. Since
+Resources are loaded before VRL compilation it will be possible to ensure that
+Vrl doesn't search on columns that do not exist within the datafile. Searching
+on a column that doesn't exist can prevent Vector from loading.
+
+If (when these features are implemented) the user attempts to reload a Resource
+and a column that has been Indexed no longer exists, this should prevent the
+file from being loaded. Vector will continue to use the currently loaded data.
+
+These restrictions apply only to columns that VRL is searching on. If the user
+attempts to use a field that has not been returned from the enrichment, this
+will be possible, but the value will be Null. This is largely due to the way
+that VRL works with paths into Objects.
+
+### Vrl functions
+
+Two remap functions:
+
+#### `find_row`
```

Comment:
To be clear, this is a generic function that will act on _all_ table types, right? If so, I'd prefer that we rename this to `find_table_row` (see [this comment](https://github.com/timberio/vector/pull/8400/files#r678291702) for naming) .

Reply:
> To be clear, this is a generic function that will act on all table types, right?

Correct.

Reply:
Awesome, I like that a lot.